engine/common/imagelib/img_png.c

```diff
@@ -0,0 +1,588 @@
+			for( ; i < pixel_size; i++ )
+				pixbuf[i] = raw[i] + ( prior[i] >> 1 );
+
+			for( ; i < rowsize; i++ )
+				pixbuf[i] = raw[i] + ( ( pixbuf[i - pixel_size] + prior[i] ) >> 1 );
+			break;
+		case PNG_F_PAETH:
+			for( ; i < pixel_size; i++ )
+				pixbuf[i] = raw[i] + prior[i];
+
+			for( ; i < rowsize; i++ )
+			{
+				a = pixbuf[i - pixel_size];
+				b = prior[i];
+				c = prior[i - pixel_size];
+				p = a + b - c;
+				pa = abs( p - a );
+				pb = abs( p - b );
+				pc = abs( p - c );
+
+				pixbuf[i] = raw[i];
+
+				if( pc < pa && pc < pb )
+					pixbuf[i] += c;
+				else if( pb < pa )
+					pixbuf[i] += b;
+				else
+					pixbuf[i] += a;
+			}
+			break;
+		default:
+			Con_DPrintf( S_ERROR "Image_LoadPNG: Found unknown filter type (%s)\n", name );
+			Mem_Free( uncompressed_buffer );
+			Mem_Free( image.rgba );
+			return false;
+		}
+
+		prior = pixbuf;
+	}
+
+	// convert RGB-to-RGBA
+	if( ihdr_chunk.colortype == PNG_CT_RGB )
+	{
+		pixbuf = image.rgba;
+		raw = uncompressed_buffer;
+
+		for( y = 0; y < image.height; y++ )
+		{
+			rowend = raw + rowsize;
+			for( ; raw < rowend; raw += pixel_size )
+			{
+				*pixbuf++ = raw[0];
+				*pixbuf++ = raw[1];
+				*pixbuf++ = raw[2];
+				*pixbuf++ = 0xFF;
+			}
+		}
+	}
+
+	Mem_Free( uncompressed_buffer );
+
+	return true;
+}
+
+/*
+=============
+Image_SavePNG
+=============
+*/
+qboolean Image_SavePNG( const char *name, rgbdata_t *pix )
+{
+	int		 ret;
+	uint		 i = 0, y, outsize, pixel_size, filtered_size, idat_len;
+	uint		 ihdr_len, crc32, rowsize;
+	byte		*in, *buffer, *out, *filtered_buffer, *rowend;
+	z_stream 	 stream = {0};
+
+	if( FS_FileExists( name, false ) && !Image_CheckFlag( IL_ALLOW_OVERWRITE ))
+		return false; // already existed
+
+	if( !pix->buffer )
+		return false;
+
+	switch( pix->type )
+	{
+	case PF_RGB_24:
+		pixel_size = 3;
+		break;
+	case PF_RGBA_32:
+		pixel_size = 4;
+		break;
+	default:
+		return false;
+	}
+
+	rowsize = pix->width * pixel_size;
+
+	// get filtered image size
+	filtered_size = ( rowsize + 1 ) * pix->height;
+
+	out = filtered_buffer = Mem_Malloc( host.imagepool, filtered_size );
+	in = pix->buffer;
+
+	// apply adaptive filter to image
+	for( y = 0; y < pix->height; y++ )
+	{
+		*out++ = PNG_F_NONE;
+		rowend = in + rowsize;
+		for( ; in < rowend; )
+		{
+			*out++ = *in++;
+		}
+	}
+
+	// get IHDR chunk length
+	ihdr_len = sizeof( png_ihdr_t );
+
+	// predict IDAT chunk length
+	idat_len = deflateBound( NULL, filtered_size );
+
+	// calculate PNG filesize
+	outsize = sizeof( png_sign );
+	outsize += sizeof( ihdr_len );
+	outsize += sizeof( ihdr_sign );
+	outsize += sizeof( png_ihdr_t );
+	outsize += sizeof( crc32 );
+	outsize += sizeof( idat_len );
+	outsize += sizeof( idat_sign );
+	outsize += idat_len;
+	outsize += sizeof( crc32 );
+	outsize += sizeof( uint );
+	outsize += sizeof( iend_sign );
+	outsize += sizeof( crc32 );
+
+	buffer = (byte *)Mem_Calloc( host.imagepool, outsize );
+
+	// write PNG header
+	buffer[0] = 0x89;
```

Comment:
I think, you can `memcpy` png_sign here, as well as other signatures(IDAT, IHDR) or fill structure and then copy it at once.